### ISSUE TITLE
Map remote: Can't map two URLs with partial path matching

### ISSUE THREAD

**Issue by cameroncooke** [ProxymanApp/Proxyman#938]
I want to map the following URL to the target URL while preserving **some** of the path:

```
api.example.com/v1/* -> `example.com/service/api/*
```

I want to use the path from the `*` onwards as the entire path to be used as the input on the target path.

Example matches would be:

```
https://api.example.com/v1/test/thing -> https://example.com/service/api/test/thing
https://api.example.com/v1/something/else -> https://example.com/service/api/something/else
```

I can't see a way to do this currently because the Map Remote tool allows you to replace the entire path but not match part of the path.

I thought about using the RegEx feature and using capture groups but I don't think you can use the replacements patterns in the "Map To -> Path" field.

### Proxyman version? (Ex. Proxyman 1.4.3)

2.29.0 (22900)

### macOS Version? (Ex. mac 10.14)

12.0 Beta (21A5284e)

### Expected behavior

To be able to use the wildcard part of the source path as the target path or support regex capture groups from the source URL in the target path field.

### Screenshots (optional)

<img width="988" alt="Screenshot 2021-07-15 at 11 42 28" src="https://user-images.githubusercontent.com/630601/125775561-8eb27310-1efa-432d-82f4-3250b2b56819.png">

**Comment by chaoyangnz** [comment]
yep, it is useless when we cannot capture the group and do replacement

**Comment by NghiaTranUIT** [comment]
It's straightforward to achieve it if you're using the Scripting tool @chaoyangnz 

Please check out this snippet code: https://docs.proxyman.io/scripting/snippet-code#change-request-destination-scheme-host-port-path

Just use some if statement and write your own logic with simple Javascript code 👍 

If you'd like using Regex, it's possible too: https://docs.proxyman.io/scripting/snippet-code#regex

I'm reluctant to implement this feature because it might introduce many edge cases that I could not cover yet.

**Comment by cameroncooke** [comment]
To be fair, it’s not that simple. Requires learning JavaScript and
understanding the API.

Being able to just pattern match is far easier. I’m not a fan of in-app
scripting environments. Seems like a way to say to your end-users who need
a feature “write your own”. The thing is we pay for this product already.
If I need to mock APIs I’ll use Postman or any other dedicated tool.

If adding this as a new feature could introduce edge cases god knows what
scripting this could do?

How comes Charles has no problem doing this?

If this app has a robust test suite as I would expect then adding this
feature should be fairly easy to validate.

**Comment by NghiaTranUIT** [comment]
it makes sense that Proxyman should implement this feature, which is supported by Charles Proxy. I will put it in our road map 👍

**Comment by cameroncooke** [comment]
Thank you for being open to being challenged, it's quite refreshing. Keep up the good work!

**Comment by chaoyangnz** [comment]
> It's straightforward to achieve it if you're using the Scripting tool @chaoyangnz
> 
> Please check out this snippet code: https://docs.proxyman.io/scripting/snippet-code#change-request-destination-scheme-host-port-path
> 
> Just use some if statement and write your own logic with simple Javascript code 👍
> 
> If you'd like using Regex, it's possible too: https://docs.proxyman.io/scripting/snippet-code#regex
> 
> I'm reluctant to implement this feature because it might introduce many edge cases that I could not cover yet.

Definitely I know scripting can achieve.
It’s nice to enhance scripting tool, I am not sure which JS engine is used, I cannot use lots of JS objects which is normally in browser and nodejs. When I need to implement a relatively real-life script, some libs are missing and even it provides the capability of adding libs or add-ons. It is actually not easy.

a simple request I’d like to forward a request to a server, I cannot modify request url directly, I have to parse URL into parts and assign to request host/port/path/queries, and URL parsing this basic function is not pre-defined.

**Comment by chaoyangnz** [comment]
> To be fair, it’s not that simple. Requires learning JavaScript and understanding the API. Being able to just pattern match is far easier. I’m not a fan of in-app scripting environments. Seems like a way to say to your end-users who need a feature “write your own”. The thing is we pay for this product already. If I need to mock APIs I’ll use Postman or any other dedicated tool. If adding this as a new feature could introduce edge cases god knows what scripting this could do? How comes Charles has no problem doing this? If this app has a robust test suite as I would expect then adding this feature should be fairly easy to validate.
> […](#)

I agree with you. Most of the time, we just need to map a url so as to point to another.

Scripting is flexible, but at the end of the day, it will end up with the script management: debugging, version control, sharing with peers etc.

**Comment by NghiaTranUIT** [comment]
Hey @cameroncooke  and @chaoyangnz, I've fixed this problem on this beta build: https://proxyman.s3.us-east-2.amazonaws.com/beta/Proxyman_2.35.0_Partial_group_matching_map_remote.dmg

From now, Map Remote can perform "Partial Matching". For instance,
- From: httpbin.org/v1/*
- to: http://httpbin.org:80/server/api/*

```
https://httpbin.org/v1/test/thing?data=123 -> http://httpbin.org:80/server/api/test/thing?data=123
```

At the moment, partial matching (`*`) is only applied for the Path.

<img width="1513" alt="Screen_Shot_2021-11-16_at_14_30_24" src="https://user-images.githubusercontent.com/5878421/141988371-c81aee31-9399-4b34-b832-87a22279f48c.png">

<img width="1204" alt="Screen_Shot_2021-11-16_at_14_30_10" src="https://user-images.githubusercontent.com/5878421/141988360-bc5828e0-0db7-492b-a355-853e60cee7c8.png">

**Comment by cameroncooke** [comment]
Hi,

Thanks for this, looks great, I've not had time to test it yet but it looks
exactly what we needed so thank you so much!

Cameron.
